clippy_lints/src/use_self.rs

```diff
@@ -190,20 +191,23 @@ impl<'tcx> LateLintPass<'tcx> for UseSelf {
                 }
             }
         }
-    }
-
-    fn check_body(&mut self, _: &LateContext<'_>, _: &hir::Body<'_>) {
-        // `lower_ty` cannot be called in `Body`s or it will panic (sometimes). But in bodies
-        // we can use `cx.typeck_results.node_type(..)` to get the `ty::Ty` from a `hir::Ty`.
-        // However the `node_type()` method can *only* be called in bodies.
-        if let Some(&mut StackItem::Check { ref mut in_body, .. }) = self.stack.last_mut() {
-            *in_body = in_body.saturating_add(1);
+        // Checking const items of `impl Self` blocks in which `std::thread_local!` expands into.
+        // HACK: `std::thread_local_inner` sometimes clears `macro_backtrace` of types in `check_ty`.
+        // To work around this problem, and assume that `LateLintPass` walks recursively from
+        // `check_impl_item` into `check_ty`, `check_expr`,etc. We encode the state when we enter
+        // this const item of `thread_local!` macro. Just remember to remove this state again in
+        // `check_impl_item_post`.
```

Comment:
`thread_local_inner` isn't behaving uniquely here, regular `macro_rules` would exhibit the same behaviour so we should probably skip any impl item that's from an expansion. e.g.

```rust
macro_rules! macro_inner_item {
    ($ty:ty) => {
        fn foo(_: $ty) {
            fn inner(_: $ty) {}
        }
    };
}

struct MyStruct;

impl MyStruct {
    macro_inner_item!(MyStruct);
}
```